### DISCUSSION TITLE
G34 fails with certain Z probe offsets

### DISCUSSION THREAD

**Discussion by RoP09** [mriscoc/Ender3V2S1#353] [Q&A] [answered]
I recently added a 2nd Z stepper and G34 worked fine so far (enabled the define in the FW without further changes).

Now I printed a new fan duct which has the CR Touch mount on the left side again (previously I had one with almost 0 offset in X direction).

If I now set the probe offset to the new physical dimensions
`M851 X-35.50 Y-8.50 Z-3.00`

And executed G34 again, it aborts immediately after probing the first stepper on the left:
```
Send: G34
[...]
Recv: X:145.50 Y:123.50 Z:13.00 E:0.00 Count X:11640 Y:9880 Z:5200
Recv: 
Recv: G34 Iteration: 1
Recv: Probing failed
Recv: G34 aborted.
[...]
Recv: X:145.50 Y:123.50 Z:13.00 E:0.00 Count X:11640 Y:9880 Z:5200
Recv: ok
```
M422 returns the following:
```
Send: M422
Recv:   M422 S1 X25.00 Y115.00
Recv:   M422 S2 X205.00 Y115.00
Recv: ok
```

If I use values that are closer to 0, like -10 for X offset it works fine (but of course with the wrong positions).

Mesh leveling  and bed tramming works fine with these new offsets however. I also tried maximize/center area for mesh leveling but it had no effect - same error.

In the config header file it says if no coordinates for left and right stepper are given it will take probe limits instead, but somehow this does not seem to work.

Any ideas? Do I need to manually define the values in the .h file instead?

**Comment by RoP09** [comment, marked as answer]
OK if I manually change m422 to restrict it on X Axis it works, but somehow I don't understand it cause according to the comment in the code it should automatically calculate the positions based on z probe limits:

```
Send: M422
Recv:   M422 S1 X45.00 Y115.00
Recv:   M422 S2 X169.00 Y115.00
Recv: ok
[...]
Send: G34
Recv: 
Recv: G34 Iteration: 1
[...]
Recv: 
Recv: Z2-Z1=0.03
Recv: 
Recv: G34 Iteration: 2
[...]
Recv: 
Recv: Z2-Z1=0.01
Recv: Target accuracy achieved.
Recv: Did 2 of 5
Recv: Accuracy: 0.01
```

**Comment by mriscoc** [comment]
Did you define Z_STEPPER_ALIGN_XY  ?

The error that you are getting is because the point to probe is not reachable. The nozzle could reach, but physically the probe maybe not.

```cpp
  /**
   * Define probe X and Y positions for Z1, Z2 [, Z3 [, Z4]]
   * These positions are machine-relative and do not shift with the M206 home offset!
   * If not defined, probe limits will be used.
   * Override with 'M422 S<index> X<pos> Y<pos>'.
   */
  //#define Z_STEPPER_ALIGN_XY { {  10, 190 }, { 100,  10 }, { 190, 190 } }
```

If Z_STEPPER_ALIGN_XY is not defined or M422 is not used, the printer uses to compute point to probe:
X/Y_MIN/MAX_BED, PROBING_MARGIN, MESH_MIN/MAX_X/Y, and Probe offsets.

**Reply by RoP09** [reply]
No I did not define Z_STEPPER_ALIGN_XY. 

Also if I use the default settings you give in the FW for X_MAX/bed size the G34 works, but since I have a linear rail on X axis I have to reduce the max X a bit. But it seems whenever I do that (e.g. only from 248 to 240)  and recreate the mesh area the G34 fails again.

Somehow it only works if I use M422 manually or if I set a wrong probe offset (like x: -10 instead of -35.5).

-

Set the X bed size and x max to 225 now, but G34 fails. Then I manually tried to set the S2 positions on M422 but it only works if it is below the max - probe offset, so it is not automatically calculated it seems. (also the value returned by m422 does never change when I change bed physical settings / mesh area:

<img width="376" alt="image" src="https://user-images.githubusercontent.com/82153569/189052525-6f474f01-3202-4d04-aeda-dabbaf20b43a.png">

With the above X positon in M422 it G34 works fine.

**Comment by mriscoc** [comment]
Can you give me all values for your physical settings and probe offsets?
If you reduce the X_MAX_POS, also the MESH_MAX_X must be reduced.

**Comment by RoP09** [comment]
Old (not working G34, works however with Probe X -10.0)
```
Probe  X-35.50 Y-8.50 Z-3.00
X Size 220
Y Size 230
X Max 220
Y Max 231
Z Max 250
M422 (FW defaults) S1 X25.00 Y115.00
M422 (FW defaults) S2 X205.00 Y115.00
```

New (workgin G34)
```
Probe  X-35.50 Y-8.50 Z-3.00
X Size 225
Y Size 230
X Max 225
Y Max 231
Z Max 250
M422 (changed by me) S1 X20.00 Y115.00
M422 (changed by me) S2 X189.00 Y115.00
```

**Reply by mriscoc** [reply]
XY MIN Pos are 0? Did you Maximize and Center the mesh?

M422 (FW defaults) S1 X25.00 Y115.00
M422 (FW defaults) S2 **X205.00** Y115.00   <- 205 is not a reachable point with the current `x_max_pos` and `probe_offset_x`

The maximum position for x is `x_max_pos` + `probe_offset_x` = 220 + (-35.50) = 184.5

**Reply by RoP09** [reply]
> XY MIN Pos are 0? Did you Maximize and Center the mesh?

yes

> 
> M422 (FW defaults) S1 X25.00 Y115.00 M422 (FW defaults) S2 **X205.00** Y115.00 <- 205 is not a reachable point with the current `x_max_pos` and `probe_offset_x`
> 
> The maximum position for x is `x_max_pos` + `probe_offset_x` = 220 + (-35.50) = 184.5

Yes but shouldnt that be calculated by the FW automatically? Without needing to set M422 manually?

**Reply by mriscoc** [reply]
> Yes but shouldnt that be calculated by the FW automatically? Without needing to set M422 manually?

Yes, that maximum must be automatic, I'm trying to simulate with the G30 and at least in my printer the reachable points are correct.
What is the output of C29?

**Reply by RoP09** [reply]
C29 L10.00 R189.50 F10.00 B220.00 N7 T60 ; 7x7  T=60°C

(from the currently working version)

**Reply by mriscoc** [reply]
That is ok because the maximum position for x is `x_max_pos` + `probe_offset_x` = 225 + (-35.50) = 189.5
And you set 189 for M422. Also look at the output of the C29, R189.5 is correct.

The unchanged output for M422 should be:
M422 S2 X189.50 Y115.00

**Reply by RoP09** [reply]
Yes, 189.5 is fine, but only since I set the M422 manually. It does not work without this manual change in contrast to what is written in the comment in code.
